zmq/types/src/json_message_types.rs

```diff
@@ -0,0 +1,647 @@
+//! Objects for JSON serialization and deserialization in message bodies of
+//! the ZMQ pub/sub interface. Handles JSON for the following subscriptions:
+//! * `json-full-txpool_add` (`Vec<TxPoolAdd>`)
+//! * `json-minimal-txpool_add` (`Vec<TxPoolAddMin>`)
+//! * `json-full-chain_main` (`Vec<ChainMain>`)
+//! * `json-minimal-chain_main` (`ChainMainMin`)
+//! * `json-full-miner_data` (`MinerData`)
+use cuprate_types::hex::HexBytes;
+use serde::{Deserialize, Serialize};
+
+/// ZMQ `json-full-txpool_add` packets contain an array of `TxPoolAdd`.
+///
+/// Each `TxPoolAdd` object represents a new transaction in the mempool that was
+/// not previously seen in a block. Miner coinbase transactions *are not*
+/// included. `do-not-relay` transactions *are* included. Values are not
+/// republished during a re-org.
+#[derive(Debug, Default, Clone, Serialize, Deserialize)]
+pub struct TxPoolAdd {
+    /// transaction version number. `2` indicates Ring CT (all sub-variants).
+    pub version: u8,
+    /// if not `0` and less than `500_000_000`, this is the block height when
+    /// transaction output(s) are spendable; if >= `500_000_000` this is roughly
+    /// the unix epoch block timestamp when the output(s) are spendable.
+    pub unlock_time: u64,
+    /// transaction inputs (key images) with separate rings for each input
+    pub inputs: Vec<PoolInput>,
+    /// transaction outputs
+    pub outputs: Vec<Output>,
+    /// extra data for the transaction with variable size, but limited to `1060`
+    /// bytes (`2120` hex nibbles).
+    #[serde(with = "hex::serde")]
+    pub extra: Vec<u8>,
+    /// obsolete, empty array in JSON
+    signatures: Vec<Obsolete>,
```

Comment:
```suggestion
    signatures: [Obsolete; 0],
```
In order to avoid an heap allocation.

Reply:
Yes, that's definitely better. I fixed them all. Thanks.